### ISSUE TITLE
"View" vs. "Testing/Effects" modes, rename "T6" to "V1"?

### ISSUE THREAD

**Issue by Moonbase59** [forkineye/ESPixelStick#117]
In general, I think it is  a good idea to somehow separate the "effects" from the "view". To be honest, we can’t expect a device as small as an ESP-01-based Pixelstick to support all effects a lighting controller (like QLC+ or xLights) can offer, and that’s part of the beauty of it. I’d care much more for a robust output device (which ESPixelstick _is_, by all means).

Introducing two effects myself, I think the effects stuff should somehow be separated, because it’s a mess having to edit many files just to shift the view mode to the end, you just might miss one (I did).

Maybe all effects could be held in a list inside some config file, or the like, so that both Javascript and C could reference them?

At least I’d be in favour of renaming the (currently) "T6" test (=view mode) for web sockets to something else like "V1", which would make changes considerably easier. Speaking of view modes, there also was a discussion re "zigzag" or "snake" type views for matrices (which _could_ be named "V2", "V3" or the like).

The view mode is an invaluable diagnostic tool, especially when using the ESPixelStick on a light controller.

**Comment by forkineye** [comment]
The original intent of the tab was for testing and diagnostics, but that was back when everyone used this solely for their holiday lighting displays.  That has obviously changed, with many people using the "test modes" for year round static lighting.  This became even more so once MQTT support was rolled in.  What I really want to do and what I've been brainstorming is to completely redo the way effects are handled.  Basically allowing the user to define their own run-time effects using some basic building blocks from the effects engine.  For instance, using "static" as a starting point, but having a sequence of colors that start or skip.  One could simulate a traditional multi-color Christmas light string and reference it as the effect "Multicolor Christmas" via MQTT.  I'll work on a json config specification that'll help explain it better.

**Comment by Moonbase59** [comment]
Sounds good (as long as it fits into the original ESPixelStick hardware).

I currently play with the EffectEngine in order to introduce __effectReversed_ and __effectMirrored_ since I feel that might be needed depending on the mounting situation and length of strip.

Also, I’m planning a realistic "fire" effect (mainly for vertical displays), much like the _Fire2012WithPalette_ from the FastLED examples. (I miss my open fireplace, sigh …)

Would it make sense to prepare that for a PR or are you already heavily working on it?

**Comment by forkineye** [comment]
It'll be a while before I can work on it, so feel free to do a PR for it. Thanks!

**Comment by Moonbase59** [comment]
Ok, working on it, I’ll let you know. :-)

**Comment by Moonbase59** [comment]
In general, for things like reversed and mirrored mode, would you prefer

* _action verbs_, like "mirror" and "reverse", or
* _adjectives_, like "mirrored" and "reversed"?

These may later on appear not only in the code but also in MQTT messages or config files, so I think this should be decided on.

Example JSON MQTT message:
```json
{"state": "ON", "color":{"r":255, "g":0, "b":0}, "effect":"Chase", "reversed":true, "mirrored":true}
```
vs.
```json
{"state": "ON", "color":{"r":255, "g":0, "b":0}, "effect":"Chase", "reverse":true, "mirror":true}
```

**Comment by penfold42** [comment]
Verb is my vote

**Comment by forkineye** [comment]
I think verb is the way to go.  I envision this mostly staying in the config file (effects.json?) on the ESPixelStick as well.  That way it would be easy for people to share effects regardless of the home automation solution, as long as they can support our basic MQTT JSON configuration utilizing the effect name.  There's another use case as well where people just use the web ui and use the test modes for off season lighting.  I would think the next step would be deciding the basic effects that we want in the effects engine core, and the verbs required to modify them.  Thanks for taking this on!

**Comment by Jchuchla** [comment]
I'm joining the conversation.  I had a ticket 103 which was a feature request to be able to send a string of channel values via MQTT to set colors in the output buffer.  Shelby closed the ticket because it would be covered by the work here.  
My specific use case is for my landscape floods.  I currently use MQTT to set them to static colors for various holidays.  But I'm limited to a single color for the whole string.  I'd like to be able to set an alternating pattern of red/white/blue or similar.  I could see this feature also being used for strings of pixels outlining a house or patio or other similar applications.

**Comment by Moonbase59** [comment]
@forkineye @penfold42: Okay, verbs it is!
Reverse and Mirror work here but I still have some issues with the Javascript and probably need some help here:
* Effect stuff is scattered throughout, I’d like to consolidate it into one wsUpdate function. How would I get at the currently shown effect’s data (r,g,b,reverse,mirror)?
* When selecting another effect, it doesn’t start automatically. This means you have to click any option (color, reverse, mirror) to get it going. I tried to send empty JSON messages like `T2{}` in `hideShowTestSections()` (to simply start the effect) but failed miserably.
* When changing to another effect, the engine keeps the last color, reverse and mirror status but doesn’t reflect this in the Web GUI when initially entering the effect’s "page".

_Reverse_ and _Mirror_ can work with any effect, but it’s up to the effect to utilize these, see "Chase" and "Rainbow". I don’t simply reverse the whole strip (like I could in setPixel) but leave this to the effect which in turn _could_ want to also mirror (by calculating only half the pixels and setting two pixels) and thus both halves must be mirrored instead of the whole strip.

**Does it make sense to branch this as a feature and make a PR so you could cross-check?**

@Jchuchla: I’m currently only adding basic functionality like "Reverse" (for different mounting situations) and "Mirror" (mainly for longer strips, the effects begin in the center). Before we start adding new effects I think the main engine must be totally robust (and configurable/extensible). Shelby has some ideas which he’ll probably introduce later on. I think we all might have some input once the idea is open for discussion.

In general, I feel we shouldn’t overdo it with the effects—from a certain point onwards, it’s _much_ more flexible to use the original DMX functionality and do it using a light controller like QLC+ or xLights.
I still agree that _some_ (MQTT-controllable) effects are great to have for all-year home lighting. ;-) but I actually miss lots of functions FastLED offers in that respect (palettes, HSV calculations, timed fading, whatnot).

**Comment by forkineye** [comment]
In regards to things be scattered about, I need to go through and rework the whole way the websocket stuff is being handled.  It's way to cumbersome adding functionality in its current state.  I have some ideas I'll prototype this weekend and see if I can't get something implemented next week.  All your stuff should fall into place though.  Thanks!

**Comment by Moonbase59** [comment]
My pleasure! Yeah, I was thinking along the same lines but I didn’t want to start and mess up things because you said you had some plans.

Apart from restructuring the web sockets server and the effects, do you think we could probably use at least part of FastLED, maybe without using their pixel driver? They have some great concepts that make effects really easy to program, i.e., the "compact" palettes, HSV calculation and so forth.

Anyway, let me know if I can help with anything.

**Comment by forkineye** [comment]
I changed effects to initialize defaults in the constructor with the latest commit.  Made it a little cleaner to persist states via MQTT.  I take a look at FastLED, know they have some good algorithms in there for sure.

**Comment by penfold42** [comment]
Finally got around to working on ESPixelBoard and starting to cherry pick commits to keep in sync.

Renaming T6 to V1 would make my ongoing life much easier

**Comment by forkineye** [comment]
Thinking something along the lines of this.  Moving view stream off to another tab to include other diagnostic functions and dedicating the "Testing" tab to effects.  This will remove view and disabled from the effects list will contain only effects.  What you all think?
![image](https://user-images.githubusercontent.com/4047923/50858206-c173f880-135d-11e9-974c-7057a9c7cbb8.png)

**Comment by Jchuchla** [comment]
Looks reasonable. Is this also where we’d see a check box to enable toggling test modes from the gpio/prog button?

**Comment by forkineye** [comment]
@Jchuchla We could have an option to enable it here, but I was thinking of just adding it as a default behavior.  The GPIO used would be a compile time option (GPIO0 / FLASH by default).  The initial effect would be whatever was saved last (Solid by default) and short presses would go through the list.

**Comment by penfold42** [comment]
@forkineye - yep agree.
I’ve already moved stream view in my fork.
You want me to cherry pick a pull request with all the untangling ?

**Comment by forkineye** [comment]
@penfold42 Let me get this pushed up first, may make a rats nest.

**Comment by penfold42** [comment]
Cool - “Disabled” needs to stay in the effects list (and probably rename to “Off”)

**Comment by forkineye** [comment]
@penfold42 What's the issue with disabling via the toggle "Enable now"? I guess slot 0 in the effects list could always be the disabled entry if needed.

**Comment by penfold42** [comment]
Sorry - my eyes glossed over that new checkbox!

**Comment by forkineye** [comment]
@penfold42 The working copy is here if you want to separate the view / effect stuff - https://github.com/forkineye/ESPixelStick/tree/effects_cleanup.  I didn't do much with it so the new checkbox isn't hooked it. Been playing with MQTT discovery in HA which seems to be working.

**Comment by penfold42** [comment]
Live view is now on the diag tab here:
https://github.com/penfold42/ESPixelBoard/tree/streamview_tab

**Comment by forkineye** [comment]
Everything is merged into the master now.  I removed the "Enable Now" option and keep the previous behavior of selecting "Disable" from the effects list.  Thanks for moving the T/V stuff around @penfold42.
